src/pathToVictory/services/officeMatch.service.ts

```diff
@@ -24,31 +27,197 @@ export class OfficeMatchService {
+        categoryTypes = filtered
+      }
+    }
+    this.logger.debug(
+      `searchDistrictTypes: categoryTypes (post-filter) size=${categoryTypes.length}`,
+    )
 
-      this.logger.debug('miscDistricts', searchColumns)
-      return searchColumns
-    } catch (error) {
-      this.logger.error('error', error)
+    // 2d) Combine candidates in priority order: subColumns → heuristicLevelTypes → categoryTypes → remaining
+    const orderedCandidates: string[] = []
+    const pushUnique = (arr: string[]) => {
+      for (const t of arr) {
+        if (typesSet.has(t) && !orderedCandidates.includes(t)) {
+          orderedCandidates.push(t)
+        }
+      }
     }
-    return searchColumns
+    pushUnique(subColumns)
+    pushUnique(heuristicLevelTypes)
+    pushUnique(categoryTypes)
+    // add remaining API types not yet included
+    pushUnique(districtTypes)
+    this.logger.debug(
+      `searchDistrictTypes: orderedCandidates size=${orderedCandidates.length}`,
+    )
+
+    // 3) Use AI to select the top 1-5 best matching district type columns
+    const matchResp = await this.matchSearchColumns(
+      slug,
+      orderedCandidates,
+      officeName,
+    )
+
+    let foundDistrictTypes: string[] = []
+    if (matchResp?.content) {
+      try {
+        const parsed: unknown = JSON.parse(matchResp.content)
+        if (
+          parsed &&
+          typeof parsed === 'object' &&
+          'columns' in parsed &&
+          Array.isArray((parsed as { columns?: unknown[] }).columns)
+        ) {
+          const cols = (parsed as { columns: unknown[] }).columns
+          const strings = cols.filter((c) => typeof c === 'string') as string[]
+          if (strings.length > 0) {
+            // Ensure all returned columns exist in API-provided list
+            foundDistrictTypes = strings.filter((c) => typesSet.has(c))
+          } else {
+            await this.slack.message(
+              {
+                body: `Received invalid response while finding district types for ${officeName}. columns: ${JSON.stringify(cols)}. Raw Content: ${matchResp.content}`,
+              },
+              SlackChannel.botDev,
+            )
+          }
+        } else {
+          await this.slack.message(
+            {
+              body: `Received invalid response while finding district types for ${officeName}. Raw Content: ${matchResp.content}`,
+            },
+            SlackChannel.botDev,
+          )
+        }
+      } catch (e) {
+        const msg = `Error parsing AI match response: ${e instanceof Error ? e.message : String(e)}`
+        this.logger.error(msg)
+      }
+    } else {
+      this.logger.debug(
+        `searchDistrictTypes: AI matching returned no content for office="${officeName}"`,
+      )
+    }
+
+    this.logger.debug(
+      `searchDistrictTypes: returning ${foundDistrictTypes.length} matched district types: ${JSON.stringify(foundDistrictTypes)}`,
+    )
+    return foundDistrictTypes
   }
```

Comment:
> Total Decision Points: Approximately 26-28 decision points

> Cyclomatic Complexity: 27-29

> This is a very high complexity score. Functions with complexity over 10 are generally considered difficult to test and maintain. This function would benefit from refactoring into smaller, more focused functions.

😳 29!

Can we not break this up? Are we really ok w/ this level of unmaintainability?